gossip/ordering/event_buffer.go

```diff
@@ -86,19 +95,15 @@ func (buf *EventBuffer) pushEvent(e *event) {
 	}
 
 	// now child events may become complete, check it again
-	for _, childID := range buf.incompletes.Keys() {
-		child, _ := buf.incompletes.Peek(childID) // without updating the "recently used"-ness of the key
-		if child == nil {
-			continue
-		}
-		for _, parent := range child.(*event).Parents {
-			if parent == e.Hash() {
-				buf.pushEvent(child.(*event))
+	eHash := e.Hash()
+	buf.incompletes.Remove(eHash)
+	for _, child := range incompleteEventsList {
```

Comment:
check the for loops at line 100-103.

- Traversing a incompleteEventsList list (100) and add something into the list (103).
- Recursive call (line 103). Check infinite loop?

Reply:
On line 103, we don't alter incompleteEventsList for performance reasons (incompleteEventsList is an array, which is fast to iterate, but hard to erase elements from beginning).

incompleteEventsList is calculated only once within one EventBuffer.PushEvent call (line 52)

Even if we did change incompleteEventsList every time, then still we wouldn't have an infinite loop, because during re-pushing of an existing event, the number of incomplete events may only decrease but not increase

Reply:
I see. Agree. So incompleteEventsList is an immutable copy of buf.incompletes, although buf.incompletes may be modified (line 66).

Btw, 
- incompleteEventsList may be not aware of newly completed event hash (line 99). 
- if incompleteEventsList is immutable, then it has to traverse the list every time (line 100).
- passing the array incompleteEventsList around recursively. Performance?